src/v/kafka/server/handlers/txn_offset_commit.cc

```diff
@@ -22,14 +22,137 @@
 
 namespace kafka {
 
+struct txn_offset_commit_ctx {
+    request_context rctx;
+    txn_offset_commit_request request;
+    ss::smp_service_group ssg;
+
+    // topic partitions found not to existent prior to processing. responses for
+    // these are patched back into the final response after processing.
+    absl::flat_hash_map<
+      model::topic,
+      std::vector<txn_offset_commit_response_partition>>
+      nonexistent_tps;
+
+    txn_offset_commit_ctx(
+      request_context&& rctx,
+      txn_offset_commit_request&& request,
+      ss::smp_service_group ssg)
+      : rctx(std::move(rctx))
+      , request(std::move(request))
+      , ssg(ssg) {}
+};
+
 template<>
 ss::future<response_ptr> txn_offset_commit_handler::handle(
-  request_context ctx, [[maybe_unused]] ss::smp_service_group g) {
-    return ss::do_with(std::move(ctx), [](request_context& ctx) {
-        txn_offset_commit_request request;
-        request.decode(ctx.reader(), ctx.header().version);
-        return ss::make_exception_future<response_ptr>(std::runtime_error(
-          fmt::format("Unsupported API {}", ctx.header().key)));
+  request_context ctx, ss::smp_service_group ssg) {
+    txn_offset_commit_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+
+    klog.trace("Handling request {}", request);
+
+    txn_offset_commit_ctx octx(std::move(ctx), std::move(request), ssg);
+
+    /*
+     * offset commit will operate normally on topic-partitions in the request
+     * that exist, while returning partial errors for those that do not exist.
+     * in order to deal with this we filter out nonexistent topic-partitions,
+     * and pass those that exist on to the group membership layer.
+     *
+     * TODO: the filtering is expensive for large requests. there are two things
```

Comment:
we can also take a 3rd approach which is to lazily materialize the this cache for the requests that actually make it to this particular core. i.e.: a cluster with 1M partitions might be expensive to materialize all. Given that each cores would handle a very small of that traffic.

Reply:
good idea!